src/lib.rs

```diff
@@ -753,15 +1063,42 @@ impl WriteHalf {
       return Err(WebSocketError::ConnectionClosed);
     }
 
-    if self.vectored && frame.payload.len() > self.writev_threshold {
-      frame.writev(stream).await?;
-    } else {
-      let text = frame.write(&mut self.write_buffer);
-      stream.write_all(text).await?;
-    }
+    // TODO(dgrr): Cap max payload size with a user setting?
+
+    frame.fmt_head(&mut self.buffer);
+    self.buffer.extend_from_slice(&frame.payload);
 
     Ok(())
   }
+
+  pub fn poll_flush<'a, S>(
+    &'a mut self,
+    stream: &mut S,
+    cx: &mut Context<'_>,
+  ) -> Poll<Result<(), WebSocketError>>
+  where
+    S: AsyncWrite + Unpin,
+  {
+    loop {
+      // try to flush before writing
+      if let Err(err) =
+        ready!(pin!(&mut *stream).poll_flush(cx)).map_err(Into::into)
+      {
+        break Poll::Ready(Err(err));
+      }
+
+      if self.buffer.is_empty() {
+        break Poll::Ready(Ok(()));
+      }
+
+      let written = ready!(pin!(&mut *stream).poll_write(cx, &self.buffer))?;
+      if written == 0 {
+        break Poll::Ready(Err(WebSocketError::ConnectionClosed));
+      }
+
+      self.buffer.advance(written);
+    }
```

Comment:
poll_write will also flush the buffers if necessary, so this loop ends up being simpler and more efficient in my testing.
```suggestion
    // flush the buffer
    while !self.buffer.is_empty() {
      let written = ready!(pin!(&mut *stream).poll_write(cx, &self.buffer))?;
      if written == 0 {
        return Poll::Ready(Err(WebSocketError::ConnectionClosed));
      }
      self.buffer.advance(written);
    }

    // flush the stream
    Poll::Ready(ready!(pin!(&mut *stream).poll_flush(cx)).map_err(Into::into))
```